packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ChildMessenger.sol

```diff
@@ -0,0 +1,86 @@
+ * `FxBaseRootTunnel` extended by the `Polygon_ParentMessenger` contract deployed on Polygon. This mapping ensures that
+ * the internal `_processMessageFromRoot` function is only callable indirectly by the `Polygon_ParentMessenger`.
+ */
+contract Polygon_ChildMessenger is FxBaseChildTunnel, ChildMessengerInterface, Lockable {
+    // The only child network contract that can send messages over the bridge via the messenger is the oracle spoke.
+    address public oracleSpoke;
+    // Store oracle hub address that oracle spoke can send messages to via `sendMessageToParent`.
+    address public oracleHub;
+
+    event SetOracleSpoke(address newOracleSpoke);
+    event SetOracleHub(address newOracleHub);
+    event MessageSentToParent(bytes data, address indexed targetHub);
+    event MessageReceivedFromParent(address indexed targetSpoke, bytes dataToSendToTarget);
+
+    /**
+     * @notice Construct the Polygon_ChildMessenger contract.
+     * @param _fxChild Polygon system contract deployed on Mainnet, required to construct new FxBaseRootTunnel
+     * that can send messages via native Polygon data tunnel.
+     */
+    constructor(address _fxChild) FxBaseChildTunnel(_fxChild) {}
+
+    /**
+     * @notice Set OracleSpoke address, which is the only address that can call `sendMessageToParent`.
+     * @dev Can only reset this address once.
+     * @param _oracleSpoke address of the new oracle spoke, deployed on this network.
+     */
+    function setOracleSpoke(address _oracleSpoke) public nonReentrant() {
+        require(oracleSpoke == address(0x0), "OracleSpoke already set");
+        oracleSpoke = _oracleSpoke;
+        emit SetOracleSpoke(oracleSpoke);
+    }
+
+    /**
+     * @notice Set OracleHub address, which is always the target address for messages sent from this network to
+     * the parent network.
+     * @dev Can only reset this address once.
+     * @param _oracleHub address of the new oracle hub, deployed on the parent network.
+     */
+    function setOracleHub(address _oracleHub) public nonReentrant() {
+        require(oracleHub == address(0x0), "OracleHub already set");
+        oracleHub = _oracleHub;
+        emit SetOracleHub(oracleHub);
+    }
+
+    /**
+     * @notice Sends a message to the OracleSpoke via the parent messenger and the canonical message bridge.
+     * @dev The caller must be the OracleSpoke on child network. No other contract is permissioned to call this
+     * function.
+     * @dev The L1 target, the parent messenger, must implement processMessageFromChild to consume the message.
+     * @param data data message sent to the L1 messenger. Should be an encoded function call or packed data.
+     */
+    function sendMessageToParent(bytes memory data) public override nonReentrant() {
+        require(msg.sender == oracleSpoke, "Only callable by oracleSpoke");
+        _sendMessageToRoot(abi.encode(data, oracleHub));
```

Comment:
hhmmm. is this correct? it might be an artefact of polygon that I'm not fully understanding. why are you not calling the call data something akin to the optimism one like this:
```
bytes memory dataSentToParent = abi.encodeWithSignature("processMessageFromCrossChainChild(bytes)", data);
```
where we encode the function that is going to be called on the `ParentMessenger`

Reply:
Yeah this is polygon specific. When sending a message from Polygon to Ethereum, the Polygon contract must emit `MessageSent`. Then, the user can manually relay the message when the block containing the transaction with the `MessageSent` event is checkpointed to Mainnet.

To manually relay the message, the user must call `receiveMessage` on the `Polygon_ParentMessenger` or whatever the `FxBaseRootTunnel` that is connected to this `FxBaseChildTunnel`. `receiveMessage` will then call the internal `_processMessageFromChild` method. This is why we don't need to use the `abi.encodeWithSignature` because we know already which internal method on the `ParentMessenger` will be called.